vars/githubWorkflowRun.groovy

```diff
@@ -0,0 +1,116 @@
+// Licensed to Elasticsearch B.V. under one or more contributor
+// license agreements. See the NOTICE file distributed with
+// this work for additional information regarding copyright
+// ownership. Elasticsearch B.V. licenses this file to you under
+// the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+/**
+
+  This step triggers workflow run on Github actions.
+
+  Example:
+     script {
+       def args = [
+           repo: "owner/repository",
+           workflow: "build.yml", 
+           ref: "main",
+           parameters: [
+             path: "filebeat"
+           ],
+           credentialsId: "github-workflow-token",
+           ghVersion: "2.1.0"]
+       def runId = githubWorkflowRun.triggerGithubActionsWorkflow(args)
+       def runInfo = githubWorkflowRun.getWorkflowRun(args + [runId: runId]) 
+   }
+*/
+
+import com.cloudbees.groovy.cps.NonCPS
+
+def call(Map args = [:]) {
+    // time in minutes to wait before build completed
+    def buildTimeLimit = args.get('buildTimeLimit', 30)
+    def runId = triggerGithubActionsWorkflow(args)
+    def startDate = new Date()
+    // the following loop intended to wait till triggered run completed
+    // or time out specified by buildTimeLimit is reached
+    while(startDate.time  + buildTimeLimit*60000 -  new Date().time > 0) {
+        def runInfo = getWorkflowRun(args + [runId: runId])
+        if (runInfo.message == "Not Found") error("Triggered workflow run but failed to get run info")
+        if (runInfo.status == "completed") return runInfo
+        sleep(buildTimeLimit > 10 ? 300 : 60)
+    }
+    error("Build time out")
+}
+
+def triggerGithubActionsWorkflow(Map args = [:]) {
+    if (!args.workflow) error('triggerGithubActionsWorkflow: workflow parameter is required.')
+    def ref = args.get("ref", "master")
+    def runner = args.get("runner", "ubuntu-latest")
+    def lookupId = args.get("lookupId", "${ref}-${new Date().getTime()}-${env.BUILD_ID}")
+    def repo = args.get("repo", "${env.ORG_NAME}/${env.REPO_NAME}")
+    def parameters = args.get("parameters", [:])
+    def inputs = (parameters + [id: lookupId, runner: runner]).collect{ "${it}" }
+    gh(ghDefaultArgs(args) + [command: "workflow run ${args.workflow}", 
+        forceInstallation: true, flags: [repo: repo, ref: ref, field: inputs]])
+    sleep(30)
```

Comment:
Is it possible for lookupForRunId() to poll instead of having this arbitrary sleep value here?

Reply:
Yes, it's possible but each unsuccessful attempt will take 101 api calls to github. I think it also make sense because will also to make it more resilient to network problems.

Reply:
Thanks, @nxei . That explanation makes sense.